src/Denudation/Abstract.jl

```diff
@@ -24,7 +24,12 @@ function denudation(input)
             end
 
             if water_depth[idx] >= 0
-                (denudation_mass[idx]) = denudation(input.box, input.denudation, water_depth[idx], slope[idx], input.facies[f])
+                denu_result = denudation(input.box, input.denudation, water_depth[idx], slope[idx], input.facies[f])
+                if denu_result !== nothing
+                    denudation_mass = Array{typeof(denu_result)}(undef, size(state.ca)...)
```

Comment:
Now you're creating a new Array filled with undefined values at every iteration. Please follow my previously sketched solution.

Reply:
I tried to now make the denudation function running on all grids by down the `for` loop on level (i.e., encapsulate for loop within individual `denudation` function). By doing so, I can allocate storage within each `denudation` function and no need to use ```Union{Array{typeof(1.0u"m/kyr"),2}, Nothing}``` and avoiding generating redundant arrays.  I didn't do the broadcasting way because I haven't find an efficient way to deal with the cell-specific facies. However, the `denudate` function in Abstract.jl was now extremely shortened and I am not sure whether it is still should be there. I kept it now because try to keep consistency with `redistribution`. The codes itself is runnable, but I have to upgrade tests yet because the tests were still designed to work on pixel-wise `denudation` function. If this way works, I would then upgrade the test.

Reply:
My suggestion is to correct the tests. Please also remove the junk code in https://github.com/MindTheGap-ERC/CarboKitten.jl/blob/f28c9ed899fcfaea84e619b15fcb5343d93ee6aa/test/DenudationSpec.jl#L57-L63